src/app/public/modules/dock/dock.service.ts

```diff
@@ -0,0 +1,208 @@
+import {
+  ComponentRef,
+  ElementRef,
+  Injectable,
+  Renderer2,
+  RendererFactory2
+} from '@angular/core';
+
+import {
+  MutationObserverService,
+  SkyDynamicComponentService
+} from '@skyux/core';
+
+import {
+  Observable
+} from 'rxjs/Observable';
+
+import {
+  Subject
+} from 'rxjs/Subject';
+
+import 'rxjs/add/observable/fromEvent';
+
+import 'rxjs/add/operator/debounceTime';
+
+import {
+  SkyDockComponent
+} from './dock.component';
+
+import {
+  SkyDockingOptions
+} from './docking-options';
+
+const ATTR_STACK_ORDER = 'data-sky-dock-stack-order';
+
+/**
+ * This service docks elements to specific areas on the page.
+ */
+@Injectable()
+export class SkyDockService {
+
+  private get bottomDockElement(): HTMLElement {
+    return this.bottomDockRef.instance.elementRef.nativeElement as HTMLElement;
+  }
+
+  private bottomDockHeight: number;
+
+  private bottomDockObserver: MutationObserver;
+
+  private bottomDockRef: ComponentRef<SkyDockComponent>;
+
+  private bottomDockStyleElement: HTMLStyleElement;
+
+  private ngUnsubscribe: Subject<void>;
+
+  private renderer: Renderer2;
+
+  constructor(
+    private mutationService: MutationObserverService,
+    private dynamicComponentService: SkyDynamicComponentService,
+    rendererFactory: RendererFactory2
+  ) {
+    this.renderer = rendererFactory.createRenderer(undefined, undefined);
+  }
+
+  /**
+   * Docks an element to the bottom of the page.
+   * @param elementRef The element to dock.
+   * @param options Options that affect the docking action.
+   */
+  public dockToBottom(elementRef: ElementRef, options: SkyDockingOptions = {}): void {
+    if (!this.bottomDockRef) {
+      this.createBottomDock();
+      this.watchDomChanges();
+    }
+
+    if (isNaN(options.stackOrder)) {
+      options.stackOrder = this.getHighestStackOrder() + 1;
+    }
+
+    const subjectElement = elementRef.nativeElement;
+    const insertBeforeElement = this.findInsertSibling(options.stackOrder);
+    const bottomDockElement = this.bottomDockElement;
+
+    this.renderer.setAttribute(subjectElement, ATTR_STACK_ORDER, `${options.stackOrder}`);
```

Comment:
Just curious - why did you decide to put properties on the DOM element instead of keep an array inside the service with this information?

Reply:
Since the elements themselves will be "physically" positioned into the desired order, it makes it a ton easier to loop through and determine the highest stack order if the property is married to the element (instead of doing a lookup in an array). Also, since the DOM is already storing the order for us, it seemed redundant to create an array of elements that may or may not be identical in order to what's already visible on the screen.

Reply:
> since the DOM is already storing the order for us...
word.